// @libp2p/upnp-nat/dist/src/gateway-finder.js
import { TypedEventEmitter, start, stop } from '@libp2p/interface';
import { repeatingTask } from '@libp2p/utils/repeating-task';
import { DEFAULT_GATEWAY_SEARCH_INTERVAL, DEFAULT_GATEWAY_SEARCH_TIMEOUT } from './constants.js';
export class GatewayFinder extends TypedEventEmitter {
    log;
    gateways;
    findGateways;
    portMappingClient;
    started;
    constructor(components, init) {
        super();
        this.log = components.logger.forComponent('libp2p:upnp-nat');
        this.portMappingClient = init.portMappingClient;
        this.started = false;
        this.gateways = [];
        // every five minutes, search for network gateways for one minute
        this.findGateways = repeatingTask(async (options) => {
            for await (const gateway of this.portMappingClient.findGateways({
                ...options,
                searchInterval: 10000
            })) {
                if (this.gateways.some(g => {
                    return g.id === gateway.id && g.family === gateway.family;
                })) {
                    // already seen this gateway
                    continue;
                }
                this.gateways.push(gateway);
                this.safeDispatchEvent('gateway', {
                    detail: gateway
                });
            }
        }, DEFAULT_GATEWAY_SEARCH_INTERVAL, {
            runImmediately: true,
            timeout: DEFAULT_GATEWAY_SEARCH_TIMEOUT
        });
    }
    async start() {
        if (this.started) {
            return;
        }
        this.started = true;
        await start(this.findGateways);
    }
    /**
     * Stops the NAT manager
     */
    async stop() {
        await stop(this.findGateways);
        this.started = false;
    }
}
